samcli/lib/init/template_modifiers/application_insights_template_modifier.py

```diff
@@ -0,0 +1,94 @@
+Class used to parse and update template when application-insights is enabled
+"""
+import logging
+from typing import Any
+from samcli.lib.init.template_modifiers.cli_template_modifier import TemplateModifier
+
+LOG = logging.getLogger(__name__)
+
+
+class ApplicationInsightsTemplateModifier(TemplateModifier):
+    import ruamel.yaml
+    from ruamel.yaml import YAML
+    from ruamel.yaml.comments import CommentedMap
+
+    TYPE_KEY = "Type"
+    RESOURCES_KEY = "Resources"
+    PROPERTIES_KEY = "Properties"
+    NAME_KEY = "Name"
+    RESOURCE_QUERY_KEY = "ResourceQuery"
+    RESOURCE_GROUP_NAME_KEY = "ResourceGroupName"
+    AUTO_CONFIG_ENABLED_KEY = "AutoConfigurationEnabled"
+    DEPENDS_ON_KEY = "DependsOn"
+    RESOURCE_GROUP_TYPE = "AWS::ResourceGroups::Group"
+    CFN_STACK_TYPE = "CLOUDFORMATION_STACK_1_0"
+    APPLICATION_INSIGHTS_TYPE = "AWS::ApplicationInsights::Application"
+    RESOURCE_GROUP_REF = "ApplicationResourceGroup"
+    APPLICATION_INSIGHTS_REF = "ApplicationInsightsMonitoring"
+    AUTO_CONFIG_VALUE = "true"
+    RESOURCE_GROUP_NAME = {"Fn::Join": ["", ["ApplicationInsights-SAM-", {"Ref": "AWS::StackName"}]]}
+
+    # set ignore aliases to true. This configuration avoids usage yaml aliases which is not parsed by CloudFormation.
+    class NonAliasingRTRepresenter(ruamel.yaml.representer.RoundTripRepresenter):
+        def ignore_aliases(self, data):
+            return True
+
+    def __init__(self, location):
+        self.yaml = ApplicationInsightsTemplateModifier.YAML()
+        self.yaml.Representer = ApplicationInsightsTemplateModifier.NonAliasingRTRepresenter
+        super().__init__(location)
+
+    def _get_template(self) -> Any:
+        with open(self.template_location) as file:
+            return self.yaml.load(file)
+
+    def _update_template_fields(self):
+        """
+        Add new resources to SAM template
+        """
+        self._add_app_insights_monitoring_section()
+
+    def _add_app_insights_monitoring_section(self):
+        resourceGroup = {
+            self.TYPE_KEY: self.RESOURCE_GROUP_TYPE,
+            self.PROPERTIES_KEY: {
+                self.NAME_KEY: self.RESOURCE_GROUP_NAME,
+                self.RESOURCE_QUERY_KEY: {self.TYPE_KEY: self.CFN_STACK_TYPE},
+            },
+        }
+
+        appInsightsApplication = {
+            self.TYPE_KEY: self.APPLICATION_INSIGHTS_TYPE,
+            self.PROPERTIES_KEY: {
+                self.RESOURCE_GROUP_NAME_KEY: self.RESOURCE_GROUP_NAME,
+                self.AUTO_CONFIG_ENABLED_KEY: self.AUTO_CONFIG_VALUE,
+            },
+            self.DEPENDS_ON_KEY: self.RESOURCE_GROUP_REF,
+        }
+
+        self.template[self.RESOURCES_KEY][self.RESOURCE_GROUP_REF] = \
+            ApplicationInsightsTemplateModifier.CommentedMap(resourceGroup)
+        self.template[self.RESOURCES_KEY][self.APPLICATION_INSIGHTS_REF] = \
+            ApplicationInsightsTemplateModifier.CommentedMap(appInsightsApplication)
+        LOG.info(self.template)
```

Comment:
Why do we need to log the template?

Reply:
Just for debugging. No specific reason

Reply:
Can we remove it? using `LOG.info` will print the template to customers' console when they run the command.